feat(ivy): add support for windows concrete types for paths

alan-agius4 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: Fixes https://github.com/angular/angular/issues/28754


## What is the new behavior?
This commit introduces support for the windows paths in the new concrete types mechanism that was introduced in this PR https://github.com/angular/angular/pull/28523

Normalized posix paths that start with either a `/` or `C:/` are considered to be an absolute path.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

devversion commented on packages/compiler-cli/src/ngtsc/path/src/util.ts:
```diff
@@ -24,3 +25,10 @@ export function normalizeSeparators(path: string): string {
 }
+
+/**
+ * Returns true if the POSIX path is an absolute path.
```
Logic-wise this makes sense, but for the comment. As per POSIX, an absolute path starts with `/` or it's not considered an absolute path ([see definition](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap04.html#tag_04_11)).

I wonder if we should just avoid the word `posix` completely because technically `normalizeSeparators` is not converting the path into a real posix path (e.g. `C:\\foo` should be `/c/foo`).

alan-agius4 replied:
Yeah that make sense, not sure, though what we'd want, ie: use proper posix or just normalize the same way that `tsc` does.

alan-agius4 replied:
Yeah that make sense, not sure, though what we'd want, ie: use proper posix or just normalize the same way that `tsc` does.

Though changing this a lot without a Windows CI would be hard not to have further regressions.

devversion replied:
Yeah, not necessarily needs to be part of that PR, but we should consider what way we want to go (e.g. either posix; or just some logic that handles kind of both)

Right now `normalizeSeparators` still says it creates `posix` paths. This is ambiguous and error-prone (the reason we have this PR 😁)

alan-agius4 replied:
ill change that comment :P

filipesilva commented on packages/compiler-cli/src/ngtsc/path/test/types_spec.ts:
```diff
@@ -10,11 +10,13 @@ import {AbsoluteFsPath} from '../src/types';
        () => { expect(() => AbsoluteFsPath.from('test.txt')).toThrow(); });
     it('should convert Windows path separators to POSIX separators',
-       () => { expect(AbsoluteFsPath.from('\\foo\\test.txt')).toEqual('/foo/test.txt'); });
+       () => { expect(AbsoluteFsPath.from('C:\\foo\\test.txt')).toEqual('C:/foo/test.txt'); });
```
What about windows absolute paths forward slash, like `D:/sandbox/ivy-latest-project/node_modules/@angular/animations` (from https://github.com/angular/angular/issues/28754)? Those need to be tested too since they are produced by TS.

alan-agius4 replied:
Does paths are actually internal 
https://github.com/angular/angular/pull/28752/files#diff-e02cda89d326098aa21575759f69c414R43

petebacondarwin commented on packages/compiler-cli/src/ngtsc/path/src/util.ts:
```diff
@@ -24,3 +25,10 @@ export function normalizeSeparators(path: string): string {
+ * Returns true if the normalized path is an absolute path.
+ */
+export function isAbsolutePath(path: string): boolean {
+  return ABSOLUTE_PATH.test(path);
```
We should probably make this OS dependent too...

alan-agius4 replied:
Maybe but in linux/unix etc... `c:/` cannot be a folder name. since is it needed?

gkalpak replied:
Not sure if it would make a noticeable difference, but making this OS dependent would be faster.

alan-agius4 replied:
Kinda wondering if to use `path.isAbsolute` directly? What do you reckon?

As OS specific might be a pain to test

alan-agius4 replied:
Kinda wondering if to use `path.isAbsolute` directly? What do you reckon?

As OS specific might be a pain to test

gkalpak replied:
If `ngtsc` can always assume access to Node.js stuff (such as `path`), then `path.isAbsolute()` might be a good option.
But keep is mind, though, that `path.isAbsolute('/foo')` returns true on Windows too and I am not sure if this is desirable.

alan-agius4 replied:
@gkalpak, I made some benchmarking and actually, os specific is more slow. Here are some data over a 1000 iterations.

```
​​​​​​​​​​
​​​​​startsWith /: 0.350ms​​​​​
​​​​​​​​​​
​​​​​ABSOLUTE_PATH regExp: 0.304ms​​​​​
​​​​​​​​​​
​​​​​os specific darwin: 2.231ms​​​​​
​​​​​​​​​​
​​​​​os specific win32: 1.740ms​​​​​

```

alan-agius4 replied:
We can make it faster, by storing `os.platform` in a const but mocking it will be tricky, Specially since it’s in a nested module and not directly within the module that we are testing.

gkalpak replied:
Well, how was "os specific" implemented. I think what @petebacondarwin (and I meant) by making this line OS specific was to use a different RegExp per OS (e.g. `const ABSOLUTE_PATH = (process.platform === 'win32') ? /^[A-Za-z]:\// : /^\//;`).

But, you are right that it would require changes to make it testable. So, I think this falls out of scope for this PR.
Maybe you could just leave a comment that this is a possible future improvement.

alan-agius4 replied:
Added a todo comment, I agree with the above, but it will complicate testing and mocking quite a bit for now.

gkalpak commented on packages/compiler-cli/src/ngtsc/path/test/types_spec.ts:
```diff
@@ -10,11 +10,15 @@ import {AbsoluteFsPath} from '../src/types';
     it('should throw when creating one from a non-absolute path',
        () => { expect(() => AbsoluteFsPath.from('test.txt')).toThrow(); });
     it('should convert Windows path separators to POSIX separators',
-       () => { expect(AbsoluteFsPath.from('\\foo\\test.txt')).toEqual('/foo/test.txt'); });
```
Why not keep this one as well?

alan-agius4 replied:
Oops

gkalpak reviewed (approved):
:tada:

gkalpak reviewed (approved):
Actually, can you update the commit message to make it clear that Windows absolute paths can start with any letter (not just `C:/`). I would also add some tests with different drive letters :grin:

gkalpak reviewed (commented):
I was looking through the ngsc code some more and I think there's at least one more change that needs to be made:

[This][1] will not behave correctly, if passed paths with Windows-style separators. If `LogicalProjectPath` is always expected to have POSIX-style separators, then we should document it ([here][2] and in the [README][3]).

[1]: https://github.com/angular/angular/blob/8accc98d28249628e84136d7306fdbbe1f4caaef/packages/compiler-cli/src/ngtsc/path/src/logical.ts#L31
[2]: https://github.com/angular/angular/blob/8accc98d28249628e84136d7306fdbbe1f4caaef/packages/compiler-cli/src/ngtsc/path/src/logical.ts#L17-L20
[3]: https://github.com/angular/angular/blob/8accc98d28249628e84136d7306fdbbe1f4caaef/packages/compiler-cli/src/ngtsc/path/README.md

alan-agius4 commented:
@gkalpak, I just had a look at `LogicalProjectPath` and also reading this [readme](https://github.com/angular/angular/blob/master/packages/compiler-cli/src/ngtsc/path/README.md#logicalprojectpath).

It stats that `This path type represents a path to a file in TypeScript's logical file system`, paths separators within the tsc compiler are always posix.

Also, seeing how a `LogicalProjectPath` is obtained ex:
```
fs.logicalPathOfFile(AbsoluteFsPath.from('/test/dist/foo.ts'))
```

It should always have `posix` paths.

alan-agius4 commented:
@gkalpak, I updated the commit as requested and added a test for another drive letter

alan-agius4 committed:
feat(ivy): add support for windows concrete types for paths

This commit introduces support for the windows paths in the new concrete types mechanism that was introduced in this PR https://github.com/angular/angular/pull/28523

Normalized posix paths that start with either a `/` or `C:/` are considered to be an absolute path.

Note: `C:/` is used as a reference, as other drive letters are also supported.

Fixes #28754

gkalpak reviewed (approved):
> It states that `This path type represents a path to a file in TypeScript's logical file system`, paths separators within the tsc compiler are always posix.
> [...]
> It should always have `posix` paths, as `AbsoluteFsPath` will normalize the separators

That's great :tada:
Could you, please, add an explicit note in the docs (both the README and the `LogicalProjectPath` API docs) for people that might not know how tsc handles paths.

##
Still LGTM :smiley: 👍

alan-agius4 commented:
@gkalpak, I added the comment in the API docs, but in the readme, I felt no need as it's already mentioned.


> # The different path kinds
All paths in the type system use POSIX format (`/` separators).

devversion commented on packages/compiler-cli/src/ngtsc/path/src/util.ts:
```diff
@@ -9,9 +9,10 @@
 
 /**
- * Convert Windows-style paths to POSIX paths.
+ * Convert Windows-style seperators to POSIX seperators.
```
```suggestion
 * Convert Windows-style separators to POSIX separators.
```

alan-agius4 committed:
fixup! feat(ivy): add support for windows concrete types for paths

alan-agius4 committed:
fixup! feat(ivy): add support for windows concrete types for paths

alxhub commented on packages/compiler-cli/src/ngtsc/path/src/util.ts:
```diff
@@ -9,9 +9,10 @@
 //  TODO(alxhub): Unify this file with `util/src/path`.
 
 const TS_DTS_JS_EXTENSION = /(?:\.d)?\.ts$|\.js$/;
+const ABSOLUTE_PATH = /^([a-zA-Z]\:\/|\/)/;
```
What is TypeScript's notion of an absolute path under Windows? Does it start with `C:/`?

alan-agius4 replied:
Yes

alxhub replied:
👍

alan-agius4 commented on packages/compiler-cli/src/ngtsc/path/src/util.ts:
```diff
@@ -24,3 +25,11 @@ export function normalizeSeparators(path: string): string {
+/**
+ * Returns true if the normalized path is an absolute path.
+ */
+export function isAbsolutePath(path: string): boolean {
```
Agree, done.

alxhub commented on packages/compiler-cli/src/ngtsc/path/src/util.ts:
```diff
@@ -24,3 +27,11 @@ export function normalizeSeparators(path: string): string {
+ */
+export function isAbsolutePath(path: string): path is AbsoluteFsPath {
+  // TODO: use regExp based on OS in the future
+  return ABSOLUTE_PATH.test(path);
```
I also realized, if this is going to assert `path is AbsoluteFsPath` then it should also verify the separators are in the right format. So I actually think I was wrong earlier, and this _shouldn't_ be a type assertion function.

alan-agius4 replied:
Dropped the commit.

naveedahmed1 commented:
Any update on this? I am still receiving error with Angular 8.0.0-beta.6 and Angular CLI 8.0.0-beta.4.

alan-agius4 commented:
@naveedahmed1, this will be released in Angular 8.0.0-beta.7

naveedahmed1 commented:
@alan-agius4 now since Angular 8.0.0-beta.7 has been released I tried compiling my project with latest version of Angular.

ivy-ngcc postinstall script in package.json seem to be working fine. But it still doesn't recognizing the modules of the app and showing below error:

`ERROR in Unable to write a reference to MyDirective in D:/MyProject/src/app/shared/my-directive/my-directive.directive.ts from D:/MyProject/src/app/shared/my-directive/my-directive.module.ts`

I am using Windows, is it still the same Windows path issue or something else?

I am following the instructions mentioned in https://next.angular.io/guide/ivy

alan-agius4 commented:
@naveedahmed1, it’s best to  file a new issue with a minimal reproduction of the problem as it’s seems to be something else.

Thanks
